src/uv_truncate.c

```diff
@@ -132,6 +132,7 @@ static void uvTruncateBarrierCb(struct UvBarrier *barrier)
     if (uv->closing) {
         tracef("closing => don't truncate");
         RaftHeapFree(truncate);
+        uvMaybeFireCloseCb(uv);
```

Comment:
TBH I took it from https://github.com/canonical/raft/blob/4cbe52efe247d11ce83f59bae4a0bf7446d3b35a/src/uv_snapshot.c#L618-L624 and figured it couldn't hurt.

Reply:
It does hurt apparently https://github.com/canonical/jepsen.dqlite/actions/runs/5455855501/jobs/9949823032
```
src/uv.c:198: uvMaybeFireCloseCb: Assertion `uv->truncate_work.data == NULL' failed."
```

Will remove it.

Reply:
Hmm it's probably more subtle than that ... will investigate a bit ...

Reply:
There's still an issue in `uvBarrierClose` I think, `uv->barrier` can be set to `NULL` while there's barrier work in flight on the threadpool, leading to the assertion failure above, due to https://github.com/canonical/raft/blob/bb92ccff31c6b86487e65e4dd3171cb8e4125861/src/uv.c#L182 

In `uvBarrierClose` I think we may only set `uv->barrier` to `NULL` if `uv->truncate_work.data == NULL && snapshot_put_work == NULL` because otherwise work is being done on the thread pool. We can then rely on the callback of the ongoing work to call `UvUnblock`, set `uv->barrier = NULL` and call `uvMaybeFireCloseCb` again.
https://github.com/canonical/raft/blob/bb92ccff31c6b86487e65e4dd3171cb8e4125861/src/uv_append.c#L918-L963

Reply:
A simpler solution might just be to check for `uv->truncate_work.data != NULL` in `uvMaybeFireCloseCb` like we do for snapshot_put https://github.com/canonical/raft/blob/4cbe52efe247d11ce83f59bae4a0bf7446d3b35a/src/uv.c#L185 

I will then reintroduce the call `uvMaybeFireCloseCb` in the truncate after work callback.

Reply:
It should be fixed by this commit https://github.com/canonical/raft/pull/435/commits/8ab42ebcce626e593f947455c40dea8dc7258afc uv->barrier can only be set to NULL if there's nothing in flight on the threadpool.